hydra-node/test/Hydra/ModelSpec.hs

```diff
@@ -75,96 +72,95 @@ import Hydra.BehaviorSpec (TestHydraNode (..))
 import Hydra.Chain.Direct.Fixture (testNetworkId)
 import Hydra.ClientInput (ClientInput (..))
 import Hydra.Model (
+  Action (ObserveConfirmedTx, Wait),
   GlobalState (..),
   Nodes (Nodes, nodes),
   OffChainState (..),
+  Payment,
   WorldState (..),
+  runModel,
  )
+import qualified Hydra.Model as Model
 import Hydra.Party (Party (..), deriveParty)
 import Hydra.ServerOutput (ServerOutput (..))
-import Test.QuickCheck (Property, counterexample, forAll, property, withMaxSuccess, within)
+import Test.QuickCheck (Property, Testable, counterexample, forAll, property, withMaxSuccess, within)
+import Test.QuickCheck.DynamicLogic (DL, action, anyActions, anyActions_, forAllDL_, forAllQ, getModelStateDL, withGenQ)
 import Test.QuickCheck.Gen.Unsafe (Capture (Capture), capture)
 import Test.QuickCheck.Monadic (PropertyM, assert, monadic', monitor, run)
-import Test.QuickCheck.StateModel (Actions, runActions, stateAfter, pattern Actions)
+import Test.QuickCheck.StateModel (Actions, RunModel, runActions, stateAfter, pattern Actions)
 import Test.Util (printTrace, traceInIOSim)
-import qualified Prelude
 
 spec :: Spec
 spec = do
   prop "model generates consistent traces" $ withMaxSuccess 10000 prop_generateTraces
   prop "implementation respects model" $ forAll arbitrary prop_checkModel
+  prop "check conflict-free liveness" prop_checkConflictFreeLiveness
+
+prop_checkConflictFreeLiveness :: Property
+prop_checkConflictFreeLiveness =
+  forAllDL_ conflictFreeLiveness prop_HydraModel
+
+prop_HydraModel :: Actions WorldState -> Property
+prop_HydraModel actions = property $
+  runIOSimProp $ do
+    _ <- runActions runIt actions
+    assert True
+
+runIt :: forall s. RunModel WorldState (StateT (Nodes (IOSim s)) (IOSim s))
+runIt = runModel
+
+-- • Conflict-Free Liveness (Head):
+--
+-- In presence of a network adversary, a conflict-free execution satisfies the following condition:
+-- For any transaction tx input via (new,tx), tx ∈ T i∈[n] Ci eventually holds.
+--
+-- TODO: make the network adversarial => make the model runner interleave/delay network messages
+conflictFreeLiveness :: DL WorldState ()
+conflictFreeLiveness = do
+  anyActions_
+  st <- getModelStateDL
+  (party, payment) <- forAllQ (withGenQ (genPayment st) (const []))
+  let newTx = Model.NewTx party payment
+  action newTx
```

Comment:
Could: inline `newTx`

Reply:
Thanks. This PR is still a draft, I just refreshed it following a discussion we had with @pgrange who was interested in having a look at what we are currently working on concretely.